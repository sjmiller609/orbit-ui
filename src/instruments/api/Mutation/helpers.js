const has = (str, find) => ~str.toLowerCase().indexOf(find.toLowerCase())

export const formErrors = ({ error, errors }) => {
  if (!error) return

  let err = JSON.stringify(error).toLowerCase()
  return errors.find(e => {
    if (e.key && has(err, e.key)) return e
    else if (e.allKeys && e.allKeys.every(k => has(err, k))) return e
    else if (e.someKeys && e.someKeys.some(k => has(err, k))) return e
  })
}
