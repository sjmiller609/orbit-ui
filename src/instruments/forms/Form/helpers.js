// faltten and stringify nested keys
export const unpack = (obj, ns, hash) => {
  if (!obj) return
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keypath = ns ? `${ns}.${key}` : key
    if (value && typeof value === 'object') {
      return unpack(value, keypath, acc)
    }
    acc[keypath] = value
    return acc
  }, hash || {})
}

const roll = (keys, obj, value) => {
  const key = keys[0]
  // case 1 - not obj
  if (keys.length === 1) {
    obj[key] = value
    return obj
  }
  // case 2 - doesn't exist
  if (!obj[key]) {
    obj[key] = isNaN(keys[1]) ? {} : []
  }
  // case 3 - exists
  obj[key] = roll(
    keys.slice(1),
    Array.isArray(obj[key]) ? [...obj[key]] : { ...obj[key] },
    value
  )

  return obj
}

// convert back into obj
export const pack = (obj, type = {}) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keys = key.split('.')

    return roll(keys, acc, value)
  }, type)
}

export const packChild = ({ name, obj }) => {
  const obj2 = {}
  let isArray
  Object.entries(obj).forEach(k => {
    if (name.length < k[0].length && k[0].indexOf(name) === 0) {
      const n = k[0].slice(name.length + 1)
      // must be true for all matching
      isArray = !isNaN(n.slice(0, n.indexOf('.')))
      obj2[n] = k[1]
    }
  })
  if (Object.keys(obj2).length) return pack(obj2, isArray ? [] : {})
  return null
}

export const removeChild = ({ name, obj = {} }) => {
  const obj2 = {}
  Object.entries(obj).forEach(k => {
    if (name.length >= k[0].length || k[0].indexOf(name) !== 0) {
      obj2[k[0]] = k[1]
    }
  })
  return obj2
}
