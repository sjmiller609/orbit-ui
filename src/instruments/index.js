// NOTE: Any instruments importing other instruments must come AFTER

export { default as Row } from './flex/Row'
export { default as Box } from './flex/Box'

export { default as Card } from './cards/Card'
export { default as Table } from './cards/Table'
export { default as TableRow } from './cards/Table/TableRow'
export { default as CardForm } from './cards/CardForm'

export { default as Module } from './layouts/Module'
export { default as Header } from './layouts/Header'

export { default as Loading } from './animations/Loading'
export { default as LoadingDots } from './animations/LoadingDots'

export { default as Search } from './forms/Search'
export { default as Form } from './forms/Form'
export { default as TextField } from './forms/TextField'
export { default as Password } from './forms/Password'
export { default as Checkbox } from './forms/Checkbox'

export { default as Link } from './links/Link'
export { default as Button } from './links/Button'

export { default as Icon } from './icons/Icon'
export { default as Logo } from './icons/Logo'

export { default as P } from './text/P'
export { default as Mini } from './text/Mini'
export { default as ShowDate } from './text/ShowDate'
export { default as B } from './text/B'

export { default as H1 } from './text/H1'
export { default as H2 } from './text/H2'
export { default as H3 } from './text/H3'
export { default as H4 } from './text/H4'
export { default as H5 } from './text/H5'

export { default as Menu } from './menus/Menu'
export { default as Level1 } from './menus/Level1'

export { default as Snackbar } from './ui/Snackbar'

export { default as ContextUI } from './context/UI/Get'
export { default as ProviderUI } from './context/UI'

export { default as Query } from './api/Query'
export { default as Mutation } from './api/Mutation'
