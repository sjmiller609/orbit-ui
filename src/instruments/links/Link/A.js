import PropTypes from 'prop-types'
import React from 'react'

const A = ({ to, title, className, onClick, children, newTab }) => {
  const aProps = {
    title,
    onClick,
    className,
  }
  if (to) {
    aProps.href = to
    aProps.target = newTab === false ? null : '_blank'
  }
  return <a {...aProps}>{children}</a>
}

A.propTypes = {
  onClick: PropTypes.func,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.array,
  ]),
  newTab: PropTypes.bool,
  to: PropTypes.string,
  title: PropTypes.string,
  className: PropTypes.string,
}

export default A
