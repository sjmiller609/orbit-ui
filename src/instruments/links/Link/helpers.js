// check if url is external
export const externalUrl = url => {
  if (typeof url === 'object') return false
  const path = /^https?:\/\/|^\/\//i
  if (path.test(url) && !~url.indexOf(window.location.origin)) return true
  return
}
