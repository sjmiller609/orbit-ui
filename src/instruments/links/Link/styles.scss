@import 'src/styles/mixins.scss';

a {
  color: inherit;
  font-style: inherit;
  text-decoration: none;
  cursor: pointer;
  transition: all 250ms ease-in;
  transform: translateZ(0);
}
a * {
  cursor: pointer;
}
a:hover,
a:focus {
  transform: scale(1.03);
  border: 0;
  outline: 0;
}
a:active {
  transform: scale(0.98);
  border: 0;
  outline: 0;
}
a:visited {
  color: inherit;
}

.arrow {
  display: inline-block;
  height: 0.6em;
  margin-left: 1rem;
  transition: all 250ms ease-in;
}

.backArrow {
  display: inline-block;
  height: 0.6em;
  margin-right: 1rem;
  transform: rotate(180deg);
  transition: all 250ms ease-in;
}
