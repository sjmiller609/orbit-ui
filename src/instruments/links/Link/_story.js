import React from 'react'
import { storiesOf } from '@storybook/react'
import { Link } from 'instruments'

storiesOf('Instruments|Links', module).add('Link', () => <Link>Mock link</Link>)
