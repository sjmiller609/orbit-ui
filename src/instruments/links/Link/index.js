'use strict'
import PropTypes from 'prop-types'
import React from 'react'
import { NavLink as NavLink1, Link as NavLink2 } from 'react-router-dom'
import s from './styles.scss'
import classnames from 'classnames'
import { Icon } from 'instruments'
import { externalUrl } from './helpers'

const Link = ({
  children,
  onClick,
  style,
  to,
  className,
  newTab,
  arrow,
  backArrow,
  ...props
}) => {
  const newProps = {
    ...props,
    to,
    className: classnames(s.link, style && s[style], className),
    onClick,
    target: newTab ? '_blank' : null,
  }
  const arr = arrow && <Icon className={s.arrow} icon={arrow} />
  const backArr = backArrow && <Icon className={s.backArrow} icon={backArrow} />

  // only use ReactRouter NavLink if activeClassName (for performance)
  const NavLink = props.activeClassName ? NavLink1 : NavLink2
  /* eslint-disable react/display-name */
  /* eslint-disable no-unused-vars */
  /* eslint-disable react/prop-types */
  const Component =
    to && !externalUrl(to)
      ? NavLink
      : ({ to, newTab, className, onClick, children }) => (
          <a
            onClick={onClick}
            className={className}
            href={to}
            target={newTab === false ? null : '_blank'}>
            {children}
          </a>
        )
  return (
    <Component {...newProps}>
      <React.Fragment>
        {backArr}
        {Array.isArray(children) ? children.map(el => el) : children}
        {arr}
      </React.Fragment>
    </Component>
  )
}

Link.propTypes = {
  onClick: PropTypes.func,
  style: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.array,
  ]),
  newTab: PropTypes.bool,
  to: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  className: PropTypes.string,
  activeClassName: PropTypes.string,
  arrow: PropTypes.string,
  backArrow: PropTypes.string,
}

export default Link
