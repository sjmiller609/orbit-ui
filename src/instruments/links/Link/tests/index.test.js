import React from 'react'
import { mountWrap } from 'helpers/tests'
import { Icon } from 'instruments'
import Link from '../index'
import A from '../A'

import { externalUrl } from '../helpers'

const props = {
  onClick: jest.fn(),
  className: 'testClassName',
  style: 'blue',
  to: '/',
}

const aProps = {
  onClick: jest.fn(),
  to: '/',
  target: '_blank',
  title: 'Test',
  id: '123',
}

describe('Link', () => {
  let wrapper = mountWrap(
    <Link {...props}>
      <p>Children</p>
    </Link>
  )

  let a = mountWrap(
    <A {...aProps}>
      <p>Children</p>
    </A>
  )

  it('should be defined', () => {
    expect(Link).toBeDefined()
    expect(A).toBeDefined()
    expect(Icon).toBeDefined()
  })

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
    expect(a).toMatchSnapshot()
  })

  it('should handle arrows', () => {
    wrapper = mountWrap(
      <Link {...props} arrow="back">
        <p>Children</p>
      </Link>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should handle paths', () => {
    wrapper = mountWrap(
      <Link {...props} to={{ pathname: '/' }}>
        <p>Children</p>
      </Link>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should handle target', () => {
    wrapper = mountWrap(
      <Link {...props} newTab={true}>
        <p>Children</p>
      </Link>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('A should handle bool target', () => {
    a = mountWrap(
      <A {...aProps} target={false}>
        <p>Children</p>
      </A>
    )
    expect(a).toMatchSnapshot()
  })

  it('should handle external URL', () => {
    expect(externalUrl()).toBeFalsy()
    expect(externalUrl({ to: '/' })).toBeFalsy()
  })
})
