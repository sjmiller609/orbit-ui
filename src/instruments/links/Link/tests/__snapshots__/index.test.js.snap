// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link displays arrow icon when \`arrow\` is specified 1`] = `
<Link
  arrow="arrow"
>
  <Component
    className="link"
    target={null}
  >
    <a
      className="link"
      target="_blank"
    >
      text
      <Icon
        className="arrow"
        icon="arrow"
      >
        <img
          className="arrow"
          src={
            Object {
              "process": [Function],
            }
          }
        />
      </Icon>
    </a>
  </Component>
</Link>
`;

exports[`Link displays back arrow icon when \`backArrow\` is specified 1`] = `
<Link
  backArrow="arrow_darkBg"
>
  <Component
    className="link"
    target={null}
  >
    <a
      className="link"
      target="_blank"
    >
      <Icon
        className="backArrow"
        icon="arrow_darkBg"
      >
        <img
          className="backArrow"
          src={
            Object {
              "process": [Function],
            }
          }
        />
      </Icon>
      text
    </a>
  </Component>
</Link>
`;

exports[`Link is active route 1`] = `
<Link
  activeClassName="active"
  to="/deployments"
>
  <NavLink
    activeClassName="active"
    aria-current="page"
    className="link"
    target={null}
    to="/deployments"
  >
    <Route
      path="\\\\/deployments"
    >
      <Link
        aria-current={null}
        className="link"
        replace={false}
        target={null}
        to="/deployments"
      >
        <a
          aria-current={null}
          className="link"
          href="/deployments"
          onClick={[Function]}
          target={null}
        >
          text
        </a>
      </Link>
    </Route>
  </NavLink>
</Link>
`;

exports[`Link is external url - http 1`] = `
<Link
  to="https://www.astronomer.io"
>
  <Component
    className="link"
    target={null}
    to="https://www.astronomer.io"
  >
    <a
      className="link"
      href="https://www.astronomer.io"
      target="_blank"
    >
      text
    </a>
  </Component>
</Link>
`;

exports[`Link is external url - naked 1`] = `
<Link
  to="astronomer.io"
>
  <Component
    className="link"
    target={null}
    to="astronomer.io"
  >
    <a
      className="link"
      href="astronomer.io"
      target="_blank"
    >
      text
    </a>
  </Component>
</Link>
`;

exports[`Link is external url - www 1`] = `
<Link
  to="www.astronomer.io"
>
  <Component
    className="link"
    target={null}
    to="www.astronomer.io"
  >
    <a
      className="link"
      href="www.astronomer.io"
      target="_blank"
    >
      text
    </a>
  </Component>
</Link>
`;

exports[`Link is internal route - no slash 1`] = `
<Link
  to="deployments"
>
  <Link
    className="link"
    replace={false}
    target={null}
    to="deployments"
  >
    <a
      className="link"
      href="deployments"
      onClick={[Function]}
      target={null}
    >
      text
    </a>
  </Link>
</Link>
`;

exports[`Link is internal route - with slash 1`] = `
<Link
  to="/deployments"
>
  <Link
    className="link"
    replace={false}
    target={null}
    to="/deployments"
  >
    <a
      className="link"
      href="/deployments"
      onClick={[Function]}
      target={null}
    >
      text
    </a>
  </Link>
</Link>
`;

exports[`Link renders correctly 1`] = `
<Link>
  <Component
    className="link"
    target={null}
  >
    <a
      className="link"
      target="_blank"
    >
      text
    </a>
  </Component>
</Link>
`;
